File: graphs/cpp/Graphs/src/ConnectedComponents.cpp
//
//  ConnectedComponents.cpp
//  Graphs
//
//  Created by Peter Goldsborough on 10/01/15.
//  Copyright © 2015 Peter Goldsborough. All rights reserved.
//

#include "ConnectedComponents.hpp"
#include "Graph.hpp"

ConnectedComponents::ConnectedComponents(const Graph& graph)
: _count(0)
, _ids(graph.number_of_vertices())
{
	bitset_t visited(graph.number_of_vertices(), false);
	
	for (vertex_t vertex = 0; vertex < graph.number_of_vertices(); ++vertex)
	{
		if (! visited[vertex])
		{
			component_t component;
			
			_dfs(graph, vertex, visited, component, _count++);
			
			_components.push_back(component);
		}
	}
}

bool ConnectedComponents::connected(vertex_t first, vertex_t second) const
{
	return id(first) == id(second);
}

const std::vector<ConnectedComponents::component_t>&
ConnectedComponents::all_components() const
{
	return _components;
}

const ConnectedComponents::component_t&
ConnectedComponents::component(id_t id) const
{
	return _components[id];
}

ConnectedComponents::id_t ConnectedComponents::count() const
{
	return _count;
}

ConnectedComponents::id_t ConnectedComponents::id(vertex_t vertex) const
{
	return _ids.at(vertex);
}

ConnectedComponents::id_t ConnectedComponents::operator[](vertex_t vertex)
{
	return _ids[vertex];
}

const ConnectedComponents::id_t
ConnectedComponents::operator[](vertex_t vertex) const
{
	return _ids[vertex];
}

void ConnectedComponents::_dfs(const Graph& graph,
							   vertex_t vertex,
							   bitset_t& visited,
							   component_t& component,
							   id_t id)
{
	visited[vertex] = true;
	
	_ids[vertex] = id;
	
	component.push_back(vertex);
	
	for (auto adjacent : graph.adjacent(vertex))
	{
		if (! visited[adjacent.first])
		{
			 _dfs(graph, adjacent.first, visited, component, id);
		}
	}
}

File: graphs/java/Graphs/src/ConnectedComponents.java
/**
 * Created by petergoldsborough on 10/01/15.
 */

import java.util.ArrayList;
import java.util.BitSet;

public class ConnectedComponents
{
	public static void main(String[] args)
	{
		Graph graph = new Graph(10);

		graph.addEdge(0, 1);
		graph.addEdge(1, 2);
		graph.addEdge(1, 3);

		graph.addEdge(4, 5);
		graph.addEdge(4, 6);
		graph.addEdge(4, 7);

		graph.addEdge(8, 9);

		ConnectedComponents cc = new ConnectedComponents(graph);

		System.out.println(cc.count());
		System.out.println(cc.connected(0, 3));
		System.out.println(cc.id(5));
		System.out.println(cc.id(8));
	}

	public ConnectedComponents(Graph graph)
	{
		int numberOfVertices = graph.numberOfVertices();

		BitSet visited = new BitSet(numberOfVertices);

		count = 0;

		ids = new int[numberOfVertices];

		for (int vertex = 0; vertex < numberOfVertices; ++vertex)
		{
			if (! visited.get(vertex))
			{
				dfs(graph, vertex, visited, count++);
			}
		}
	}

	public boolean connected(int first, int second)
	{
		return id(first) == id(second);
	}

	public ArrayList<ArrayList<Integer>> allComponents()
	{
		ArrayList<ArrayList<Integer>> all = new ArrayList<>(count);

		for (int id = 0; id < count; ++id)
		{
			all.add(component(id));
		}

		return all;
	}

	public ArrayList<Integer> component(int id)
	{
		ArrayList<Integer> vertices = new ArrayList<>();

		for (int vertex = 0; vertex < ids.length; ++vertex)
		{
			if (ids[vertex] == id) vertices.add(vertex);
		}

		return vertices;
	}

	public ArrayList<Integer> singleComponentVertices()
	{
		ArrayList<Integer> members = new ArrayList<>();

		BitSet added = new BitSet(count);

		for (int vertex = 0; vertex < ids.length; ++vertex)
		{
			if (! added.get(ids[vertex]))
			{
				members.add(ids[vertex]);

				if (added.isEmpty()) break;
			}
		}

		return members;
	}

	public int count()
	{
		return count;
	}

	public int id(int vertex)
	{
		return ids[vertex];
	}

	private void dfs(Graph graph, int vertex, BitSet visited, int id)
	{
		visited.set(vertex);

		ids[vertex] = id;

		for (Graph.Edge adjacent : graph.adjacent(vertex))
		{
			if (! visited.get(adjacent.vertex))
			{
				dfs(graph, adjacent.vertex, visited, id);
			}
		}
	}

	private int count;
	private int[] ids;
}
